### PR TITLE
Multiresolution Analysis

### PR DESCRIPTION
**Pull request by grlee77**
This PR adds a function that creates projections onto wavelet subspaces. In 1D, this gives output as described by Matlab for their `MODWTMRA`. I don't think there is a 2D or nD equivalent in Matlab, but the concept is the same. 

The inverse reconstruction for this "additive" decomposition is very simple. It is just the sum of the projected coefficients. The other nice property is timeseries alignment.

One difference in API choice vs. Matlab. In Matlab, MODWTMRA takes the output of MODWT. I did not take that approach here, but instead operate on the original signal. This avoids the user potentially passing coefficients from an `swt` called without the correct options for normalization, etc.

Also, by default this analysis is performed with the stationary wavelet transforms under the hood. The projections can also be done with `wavedec/waverec` DWT functions and this is faster, but due to the lack of shift invariance the result typically looks substantially worse. 

The computational efficiency of this implementation could be improved further by skipping filtering of all of the zero-amplitude subbands during the various calls to `iswt`/`iwaverecn`, but that would require more extensive modifications to the underlying transforms.

I will attach images from two demos below.

In 1D, you can see that the outputs of MRA are temporally aligned while those from SWT are not (unless the wavelet was symmetric):
MRA

A visualization for the MRA in 2D for the cameraman image is given below:
mra2d_swt

If the above example is repeated but with `transform='dwtn'` instead of `transform='swtn'`, the result is substantially worse:
mra2d_dwt

This difference illustrates the benefit of the shift-invariance property of the SWT.

**TODO:**
- [x] add tests

### CODE REVIEW

File: pywt/_mra.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+
+    This transform does not share the variance partition property of ``swt``
+    with `norm=True`. It does however, result in coefficients that are
+    temporally aligned regardless of the symmetry of the wavelet used.
+
+    The redundancy of this transform is ``(level + 1)``.
+
+    See Also
+    --------
+    ``imra``, ``swt``
```
Comment (rgommers) [inline pywt/_mra.py]: Can you add `References` for these functions? I assume it's based on a paper?

**Comment by rgommers** [comment]
interesting

**Comment by grlee77** [comment]
Agreed on adding a references section. Matlab references Percival & Walden's "Wavelet Methods for Time Series Analysis" book for their MODWTMRA function.

There is an R package called `wmtsa` that [corresponds to that book](https://cran.r-project.org/web/packages/wmtsa/index.html). It looks like the name of the corresponding function there is `wavMRDSum`. Where MRD stands for "Multiresolution Decomposition".  The R function, like the one here, allows choosing between the DWT or SWT for the actual computation. The references given in the WMTSA docs are:

```
D. B. Percival and A. T. Walden, Wavelet Methods for Time Series Analysis, Cambridge University
Press, 2000.

T.W. Randolph and Y. Yasui, Multiscale Processing of Mass Spectrometry Data, Biometrics, 62:589–
97, 2006.
```

I did not reference either the R or Matlab code when implementing this and have not yet tested how close the result produced here gives to the output from either of those libraries. I did verify the key property that the inverse transform via simple summation reconstruction the original signal.

Not sure yet if `mra`/`imra` is the best name for this. I am thinking we want to put "additive" in the name somewhere. Technically all of the multilevel decompositions in PyWavelets are "multi-resolution" so it is probably worth being more specific. 

I am not aware of publications or implementations specifically about this additive decomposition in higher dimensions. Also, we don't strictly need both 2d and nd implementations, but I just have that here for consistency with the rest of the library.

**Comment by rgommers** [comment]
I see I missed updates here from a few months ago. This looks ready to go in; triggered CI to see if it passes on 3.10 as well.

**Comment by rgommers** [comment]
@grlee77 I didn't review in a lot of detail, but if you are happy with it I think we should merge it and take it along in the 1.2.0 release. What do you think?

**Comment by grlee77** [comment]
Yes, out of the open PRs this is the one I would most like to get into 1.2.0.

**Pull request merged**